migrations/2021_03_02_040000_change_access_tokens_add_id.php

```diff
@@ -0,0 +1,50 @@
+<?php
+
+/*
+ * This file is part of Flarum.
+ *
+ * For detailed copyright and license information, please view the
+ * LICENSE file that was distributed with this source code.
+ */
+
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Database\Schema\Builder;
+
+return [
+    'up' => function (Builder $schema) {
+        $schema->table('access_tokens', function (Blueprint $table) {
+            $table->unsignedInteger('id')->first();
+        });
+
+        $tokens = $schema->getConnection()->table('access_tokens')
+            ->cursor();
+
+        // Insert initial value for our new primary key
+        foreach ($tokens as $i => $token) {
```

Comment:
I have to say, I'm a bit worried about this one.

Because of the garbage collector bug, some forums will have accumulated tens of thousands if not more access tokens.

Should I include the equivalent of the garbage collector in there to clear all the expired tokens before the migration?

My thought was that the migration would be simpler by not going the garbage deletion, since it will happen when they start browsing the updated forum anyway.

I'm not sure if anyone has a large forum backup we could try this on.

Reply:
> I'm not sure if anyone has a large forum backup we could try this on.

My forum has 48k access tokens and I also have a local installation based on a recent backup, if necessary

Reply:
Thanks for the info!

I have modified the migrations, the new version should actually delete most of the tokens before attempting to update those that are remaining (only remember tokens will be kept).

Do you know how to run the dev-master of Flarum? That would be a great test once we merge.

I don't want to ask for a copy of the table since it's obviously sensitive data.

Reply:
> Do you know how to run the dev-master of Flarum? That would be a great test once we merge.

To be honest, last time I tried I didn't succeed, because of some Composer constraints issues IIRC. But I can try again. You know where to find me :)

Reply:
@matteocontrini actually, if you could send me a database dump of just the `access_tokens` table, but without the `token` column, that would be great! I think some tools like phpMyAdmin allow selecting columns during export.

I'm pretty sure I have a discussion with you on Discord, but I can't recall your username :see_no_evil: